tests/test_cases/test_axi/test_axi.py

```diff
@@ -0,0 +1,349 @@
+""" Copyright 2019 by Ben Coughlan <ben@liquidinstruments.com>
+
+Licensed under Revised BSD License
+All rights reserved. See LICENSE.
+"""
+
+import cocotb
+from cocotb.clock import Clock
+from cocotb.triggers import ClockCycles
+from cocotb.drivers.amba import AXI4MemoryMap
+from cocotb.drivers.amba import handshake
+from cocotb.result import ReturnValue
+from cocotb.bus import Bus
+from cocotb.monitors.amba import AXI4
+from cocotb.scoreboard import Scoreboard
+
+CLK_PERIOD = 10
+
+
+@cocotb.coroutine
+def write(clock, bus, addr, data, thread_id=0, burst_type=1):
+    if not isinstance(data, (list, tuple)):
+        data = [data]
+
+    bus.awid <= thread_id
+    bus.awaddr <= addr
+    bus.awlen <= len(data) - 1  # TODO unaligned transfers
+    bus.awsize <= 0x03  # 64 bit
+    bus.awburst <= burst_type
+    bus.awlock <= 0x00
+    bus.awcache <= 0x00
+    bus.awprot <= 0x00
+    bus.awqos <= 0x00
+
+    aw = cocotb.fork(handshake(clock, bus.awvalid, bus.awready))
+
+    @cocotb.coroutine
+    def _write_thread(clock, bus, data):
+        for i, d in enumerate(data):
+            bus.wdata <= d
+            bus.wstrb <= 0xFF
+            bus.wlast <= (1 if i == len(data)-1 else 0)
+            yield handshake(clock, bus.wvalid, bus.wready)
+            bus.wlast <= 0
+    bus.wid <= thread_id
+
+    w = cocotb.fork(_write_thread(clock, bus, data))
+
+    yield aw.join()
+    yield w.join()
+
+    @cocotb.coroutine
+    def _wait_for_bresp(clock, bus, wid):
+        while True:
+            yield handshake(clock, bus.bready, bus.bvalid)
```

Comment:
I don't really understand what's going on here. When you fork in `test_write_multithread`, you create four coroutines that are all trying to handshake on the same bus by driving the same signals

Reply:
The fork happens after the initial handshake so each transaction is initiated sequentially; awdata and wdata are technically asynchronous and are allowed to run in parallel, but the transaction isn't started until they are both complete.
The returned coroutine then waits until a response for the specific thread_id occurs on the *resp port.  So multiple transactions can be in flight simultaneously, up to max_threads; after which the start of a transaction will block until a thread opens up.